### ISSUE TITLE
Handle file firewall forbidden message and show it to the user.

### ISSUE THREAD

**Issue by SergioBertolinSG** [owncloud/android#1759]
Currently Android client is not showing the user what happened if the server blocks access by firewall.

To block the Android client add a rule in the file firewall of the server (enterprise only) like this:

![screen shot 2016-08-02 at 16 39 00](https://cloud.githubusercontent.com/assets/9048796/17332625/ad6c6eda-58cf-11e6-9c72-ee8b187215a1.png)

For example the desktop client blocked by firewall:

<img src="https://cloud.githubusercontent.com/assets/9048796/10459377/87efff92-71cf-11e5-9e9b-a4506db8ed7e.png" width="420">

cc: @davivel

**Comment by davivel** [comment]
This would be for an account as a whole, or for individual files/folder in the account?

To apply only when the account is created, or can we expect that an account is blocked by the firewall after some time of being operated, and than should be blocked?

**Comment by SergioBertolinSG** [comment]
> This would be for an account as a whole, or for individual files/folder in the account?

Both, there are filewall rules to block access from android client and rules to block upload/download of files.

The account won't be blocked, just the access from an android device. If the rule is removed from firewall, the user should be able to use android client again without problems.

**Comment by davivel** [comment]
And what's happening now? Mistery/silent errors?

**Comment by SergioBertolinSG** [comment]
With device android client firewall rule, the user can log in (which perhaps should be avoided, like the desktop client does) and when trying to download a file it keeps trying to download it without showing any errors.

When having a rule about denying upload of a specific file type using mimetype rule, android detects it as a permissions error, instead of a forbidden error (403).

**Comment by LukeOwncloud** [comment]
There are a number of actions the oC can perform which require server access:

1. update folder content (by browsing through folder or swipe down)
2. synchronize account (by choosing from three dots menu)
3. upload file
4. download file
5. setup a new account

If if understand correctly this issue is just about the first action. For the others there are error messages.


My second point: It is not really a problem with connections being blocked by a firewall, but in general if the server cannot be reached. For whatever reason: smartphone not connected, wifi has no Internet access, server down, blocked by firewall, ...


If this makes sense to you and I did not understand this issue completely wrong, you may find that this is a duplicate of https://github.com/owncloud/android/issues/688. Agreed?

**Comment by SergioBertolinSG** [comment]
Hi @LukeOwncloud, I am talking about a feature only included in enterprise version of the server called file firewall. Sorry for the confusion.

**Comment by davigonz** [comment]
@guruz @ogoffart In the login view, while checking the credentials with the rule "User Device is Android Client" enabled in firewall application, we receive a 200 OK. In desktop client you are able to receive an error in the login. Which HTTP method are you using? We use a HEAD method to root directory.

**Comment by SamuAlfageme** [comment]
@davigonz 

> In the desktop client you are able to receive an error in the login.

I'm looking into this since I'm not able to reproduce with 9.1.3. Could be a regression on our side, maybe after consuming the new (>9.0) WebDAV endpoint https://github.com/owncloud/client/issues/4007. 

To answer your question, in a 8.2.9 server, the PROPFIND to `/remote.php/webdav/` returned a 403 with the descriptive message:

```
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>OCP\SabrePluginException</s:exception>
  <s:message>Access to this resource has been forbidden by a file firewall rule.</s:message>
</d:error>
```

**Comment by guruz** [comment]
@davigonz The screenshot above is for the wizard. We don't use HEAD, usually only GET/PROPFIND. I'd need to check.

For non-wizard: For logging on, we use GET on status.php (unauthed), and then try to get authed using PROPFIND on webdav.
You can tail the access log to see.

Does this answer the question? Else i don't understand it:)

**Comment by davivel** [comment]
@guruz, thanks. The question is about our own "wizard", the log-in activity. Knowing that you don't use HEAD is good enough.

**Comment by davivel** [comment]
Maybe the time has come to

![off_with_their_heads](https://cloud.githubusercontent.com/assets/2225614/22503197/e85af9b8-e870-11e6-9133-d1211010b61a.jpg)

**Comment by davigonz** [comment]
Thank you for your comments @SamuAlfageme and @guruz, I've just fixed this issue using a 8.2.9 server, which is returning 403 forbidden as expected. 

Since it seems that servers >9.0 are not responding properly with specific 403 Forbidden firewall rules enabled, I will create a detailed issue (similar to https://github.com/owncloud/client/issues/5504) and involve the server. 

To sum up, these are the error messages we're going to handle in this issue, according to the different firewall rules:

1. **File mimetype upload & File size upload rules:**
   * Specific firewall error message in downloads notification.
   * Specific firewall error message in uploads view and notification.
  
2. **User Device is Android Client & User Group is Whatever:**
    * Forbidden authentication error message in login view.
    * Forbidden error message in uploads: we cannot show here (at the moment) a specific firewall error message because we use a HEAD method to check some stuff before starting the upload operation. In this HEAD method we get a 403 forbidden without any response body that includes a specific error.

3. **Request Time & Request Type is WebDAV**
     * Forbidden authentication error message in login view.
     * Specific firewall error message in downloads
     * Forbidden error message in uploads: the same as the uploads in 2.

4. **Request Type is WebDAV**
     * Forbidden authentication error message in login view.
     * Specific firewall error message in downloads.
     * Non specific error message in uploads: the same as the uploads in 2 .

This list can be useful for QA, @jesmrec

**Comment by davigonz** [comment]
This issue can be reviewed and pass QA @jesmrec @jabarros

**Comment by jesmrec** [comment]
Approved

CC @davigonz
